### ISSUE TITLE
exampleJava contains multiple binary copies of caj and jca

### ISSUE THREAD

**Issue by ralphlange** [epics-base/exampleJava#16]
The exampleClient sources contain copies of binary jars for caj and jca in the .../shell directory.

I don't think this is acceptable, as binaries do not belong into source structures, and the versions are not checked to match the globally required versions (in exampleJava's dependencies pulled from epics-core and epics-parent).
Building exampleClient downloads jca and caj as dependencies, so these jars should be used.

**Comment by ralphlange** [comment]
Same applies to exampleLink.

**Comment by mrkraimer** [comment]
The problem is how to specify  jca and caj in the sourceEXAMPLE files.
How can this be done  without having a local copy of jca and caj?

**Comment by ralphlange** [comment]
The maven build downloads all dependencies, so there is a local copy of jca and caj.

**Comment by mrkraimer** [comment]
Where is the local copy of jca and caj that is generated by maven?
How do I get it them in CLASSPATH? 

Look  at  exampleJava/exampleClient/shell/sourceEXAMPLE

The last three lines are:
export CLASSPATH=${CLASSPATH}:${EXAMPLECLIENT}/target/classes
export CLASSPATH=${CLASSPATH}:${EXAMPLECLIENT}/shell/caj-1.1.15.jar
export CLASSPATH=${CLASSPATH}:${EXAMPLECLIENT}/shell/jca-2.3.6.jar

jca and need to be  in the CLASSPATH.

With these line present the following works:
mrk> ./examplePvaClientGet 

With the last two lines removed the output is

mrk> ./examplePvaClientGet 
_____examplePvaClientGet starting_______
**exampleDouble** channelName PVRdouble provider pva
short way
as double 0.0
repeat short way
as double 0.0
long way
as double 0.0
**exampleDoubleArray** channelName PVRdoubleArray provider pva
short way
as doubleArray
repeat short way
as doubleArray
long way
as doubleArray
exception Failed to initialize shared CA client instance.
java.lang.RuntimeException: Failed to initialize shared CA client instance.
    at org.epics.ca.ClientFactory$ChannelProviderFactoryImpl.sharedInstance(ClientFactory.java:67)
    at org.epics.pvaccess.client.ChannelProviderRegistryFactory$ChannelProviderRegistryImpl.getProvider(ChannelProviderRegistryFactory.java:51)
    at org.epics.pvaClient.PvaClientChannel.issueConnect(PvaClientChannel.java:352)
    at org.epics.exampleClient.ExamplePvaClientGet.main(ExamplePvaClientGet.java:95)
Caused by: java.lang.NoClassDefFoundError: gov/aps/jca/event/ContextExceptionListener
    at java.lang.ClassLoader.defineClass1(Native Method)
    at java.lang.ClassLoader.defineClass(ClassLoader.java:760)
    at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
    at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
    at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
    at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
    at org.epics.ca.ClientFactory$ChannelProviderFactoryImpl.sharedInstance(ClientFac
...

**Comment by dhickin** [comment]
These are usually in the .m2 directory of your home directory, e.g.
 ~/.m2/repository/org/epics/jca/2.3.6/

The default can be overwritten in the settings.xml file in your maven installation.
E.g.
 <localRepository>/path/to/local/repo</localRepository>

There are usually instructions in the settings.xml file.

**Comment by ralphlange** [comment]
Don't try to re-invent Maven using bash.
Have a look at exampleJava/database [release/4.2] to see how to execute applications from Maven.
In that case, I was able to remove the "shell" things completely.

**Comment by mrkraimer** [comment]
I looked at exampleJava/database [release/4.2] 

I did manage to change exampleClient./pom.xml so that I can  run
org.epics.exampleClient.ExamplePvaClientGet

via mvn exec:exec

But I still have to see how the exampleClient/pom.xml has to be changed so that it works for the other exampleClient examples.
And then all the other examples under exampleJava.

This will take some work so I will start making changes in my github repository first.

I am still a novice maven user!!

**Comment by mrkraimer** [comment]
I looked at exampleJava/database [release/4.2] 

I did manage to change exampleClient./pom.xml so that I can  run
org.epics.exampleClient.ExamplePvaClientGet

via mvn exec:exec

But I still have to see how the exampleClient/pom.xml has to be changed so that it works for the other exampleClient examples.
And then all the other examples under exampleJava.

This will take some work so I will start making changes in my github repository first.

I am still a novice maven user!!

**Comment by ralphlange** [comment]
Don't do any configuration in the pom

On Linux, the script file has to be along the lines of

```
#!/usr/bin/env bash
mvn exec:exec -Dexec.args="-classpath %classpath org.epics.exampleClient.ExamplePvaClientGet $@"
```

I don't have a Linux system with Maven/Java to test on, and on Windows this is different. (I will have to add batch files for every one of the test scripts.)
Note that you have to be in the directory with the pom.xml to run this.

**Comment by mrkraimer** [comment]
I looked at exampleJava/database [release/4.2] 

I did manage to change exampleClient./pom.xml so that I can  run
org.epics.exampleClient.ExamplePvaClientGet

via mvn exec:exec

But I still have to see how the exampleClient/pom.xml has to be changed so that it works for the other exampleClient examples.
And then all the other examples under exampleJava.

This will take some work so I will start making changes in my github repository first.

I am still a novice maven user!!

**Comment by mrkraimer** [comment]
I looked at exampleJava/database [release/4.2] 

I did manage to change exampleClient./pom.xml so that I can  run
org.epics.exampleClient.ExamplePvaClientGet

via mvn exec:exec

But I still have to see how the exampleClient/pom.xml has to be changed so that it works for the other exampleClient examples.
And then all the other examples under exampleJava.

This will take some work so I will start making changes in my github repository first.

I am still a novice maven user!!

**Comment by ralphlange** [comment]
That is actually quite ugly for the example clients, because of the Maven clutter in the output.

I will think of a better way. Something like using the maven-dependency-plugin to output a classpath file which can be used on a java command line.
That would get rid of the Maven execution for each example, and thus allow the scripts to be run from any place.

**Comment by ralphlange** [comment]
Key goals:
- Use Maven to produce the classpath, as it knows best what the dependencies are.
- Allow the examples to be run in a similar fashion on Windows, without dependency to bash.

**Comment by ralphlange** [comment]
Yay - I do have a working Maven command line for Windows.
(Only for command.com - Cygwin/bash completely screws up the quoting.)

**Comment by mrkraimer** [comment]
Sorry about all the duplicate messages from me.

In my repository I did get all of  the following to work
,/examplePvaClientGet
./monitorForever -help
./monitorForever 

Here is what is being executed.

mrk> pwd
/home/epicsv4/master/exampleJava/exampleClient/shell
mrk> cat examplePvaClientGet
#!/usr/bin/env bash
cd ..
mvn exec:exec -Dexec.args="-classpath %classpath org.epics.exampleClient.ExamplePvaClientGet $@"
mrk> cat monitorForever 
#!/usr/bin/env bash
cd ..
mvn exec:exec -Dexec.args="-classpath %classpath org.epics.exampleClient.MonitorForever $@"

But it sounds like Ralph has other ideas that also solves problems with windows?

So what to do?

**Comment by mrkraimer** [comment]
Ralph said
        That is actually quite ugly for the example clients, because of the Maven clutter in the output.

The clutter can be removed by specifying -q

For example I changed monitorForever to be
#!/usr/bin/env bash
cd ..
mvn -q exec:exec -Dexec.args="-classpath %classpath org.epics.exampleClient.MonitorForever $@"

Then I get

mrk> pwd
/home/epicsv4/master/exampleJava/exampleClient/shell
mrk> ./monitorForever -help
provider channelName  debug
default
pva PVRdouble false
mrk>

**Comment by ralphlange** [comment]
Correct,  this removes the extra output. Still, the Maven run slows down the execution noticeably, that's done at every execution, while the classpath is known after the build, and does not change afterwards.

**Comment by mrkraimer** [comment]
I agree that the Maven run slows down the execution noticeably.

But I think that we have to be careful about how we have CLASSPATH defined.

If we do go with the mvn exec approach then the following also works:

mrk> pwd
/home/epicsv4/master/exampleJava/exampleClient
mrk> ./monitorForever  -help
provider channelName  debug
default
pva PVRdouble false
mrk> cat monitorForever 
mvn -q exec:exec -Dexec.args="-classpath %classpath org.epics.exampleClient.MonitorForever $@"
mrk> 

Thus no need for bash or sub-directory named shell.

**Comment by ralphlange** [comment]
Nah.  We shouldn't clutter the top level directory with arbitrary one-line scriptlets. 
I will have a look at the Maven standard directory structure and will put these where they belong.

**Comment by ralphlange** [comment]
As long as the classpath is produced by Maven, we're fine on any system.

**Comment by mrkraimer** [comment]
Let me know when You have something that works.
I suggest that you do exampleJava/database and exampleJava/exampleClient first.
If you want I can do the other examples.
I assume that you will make the changes on branch release/4.2

**Comment by ralphlange** [comment]
Almost done... stay tuned.

**Comment by ralphlange** [comment]
OK, please have a look.
The client README is not enough. It doesn't tell you what the examples do and what to expect. For the simple get (on Windows) I get some printout saying channels are not being found, and the client does not exit, only using ^C, which creates errors on the server side.

**Comment by ralphlange** [comment]
@anjohnson: the database and exampleClient parts of exampleJava now have batch files to execute the examples on Windows. I don't know if these are portable enough yet.

**Comment by mrkraimer** [comment]
I did all the other examples except rdbService (It is not being built via exampleJava/pom.xml).
I also tested all examples, except rdbService, on linux.
I think I generated the .bat files correctly  but have no way to test them.
I also updated the README.md files but not the documentation/*.html files.

Thanks Ralph!!

I like the script solution  much better than shell solution.

**Comment by anjohnson** [comment]
@ralphlange What file(s) would I have to download to be able to try and run one or more of the pre-built examples (assuming I haven't downloaded or checked out _anything_, so all I have is a Windows box with Java on it)? The point of providing pre-built software is usually so the user can run it without having to compile anything first, and maven normally doesn't come with Java anyway.

Should the scripts appear in the `-sources.jar` files? I'm not a Java programmer, but if the point of that particular collection is to contain everything needed to build this code I would expect it to be pretty much a complete copy of the git clone, which these doesn't appear to be. (This kind of ignorant question is why you really shouldn't ask me to do anything related to Java...).

I looked at your .bat files and don't see anything obviously unportable about them, but I haven't tried running them (I don't even have a git clone of any of the Java modules).

**Comment by ralphlange** [comment]
Excellent question. 
How are the Java examples supposed to be used? 
If they are supposed to show how to use the EPICS Java libraries, it would make a lot more sense to bundle the complete source tree as a Maven project - similar to C++.
A binary only example does not show much. Adding the sources without the Maven structure is like shipping C++ without the Makefiles. Nice to look at, but not very useful to show you how it's done.

**Comment by ralphlange** [comment]
Fixed by 554cbcf059af29d5ff08945c4763529b83afcf92.
